### PR TITLE
607 inspect command columns update

### PR DESCRIPTION
**Pull request by GDonRanasinghe**
Satisfies #607 .

### CODE REVIEW

File: src/csvcubed/models/sparqlresults.py [hunk 1]
```diff
@@ -582,3 +663,21 @@ def map_table_schema_properties_result(
...
+def map_is_pivoted_shape_for_measures_in_data_set(
+    sparql_results: List[ResultRow],
+) -> List[IsPivotedShapeMeasureResult]:
+    """
+    Maps the sparql query result to objects of type IsPivotedMeasureResult that are then returned.
+    """
+    is_pivoted_shape_measure_results = list(
+        map(
+            lambda result: IsPivotedShapeMeasureResult(
+                measure=str(result.asdict()["measure"]),
```
Comment (robons) [inline src/csvcubed/models/sparqlresults.py:677]: Again, can you make sure to only call `.asdict` once per result given that I think this is likely an expensive operation?

File: src/csvcubed/utils/qb/cube.py [hunk 2]
```diff
@@ -17,6 +17,7 @@
     QbObservationValue,
     QbColumnStructuralDefinition,
 )
+from csvcubed.utils.sparql_handler.sparqlmanager import CSVWShape
```
Comment (robons) [inline src/csvcubed/utils/qb/cube.py:20]: Can you store this elsewhere? I'd like to make sure that the `csvcubed build` functionality isn't dependent on `csvcubed inspect` functionality and vice versa. If you place it somewhere which isn't in one or the other we can avoid having any dependencies there.

File: src/csvcubed/utils/qb/cube.py [hunk 3]
```diff
@@ -91,3 +92,29 @@ def get_all_units(cube: Cube) -> Set[QbUnit]:
...
+            all_standard_shape
+            and not obs_val_col.structural_definition.is_pivoted_shape_observation
+        )
+
+    if all_pivoted:
+        return CSVWShape.Pivoted
+    elif all_standard_shape:
+        return CSVWShape.Standard
+    else:
+        raise TypeError("The cube cannot be in both standard and pivoted shape")
```
Comment (robons) [inline src/csvcubed/utils/qb/cube.py:120]: Could you make it clear in this error message that it's about some obs val columns being in pivoted shape and others not?

File: src/csvcubed/utils/sparql_handler/sparqlmanager.py [hunk 4]
```diff
@@ -51,6 +53,18 @@
 _logger = logging.getLogger(__name__)
 
 
+class CSVWShape(Enum):
```
Comment (robons) [inline src/csvcubed/utils/sparql_handler/sparqlmanager.py:56]: Can we call this `CubeShape` for consistency? A code list will never be in either of these shapes.

File: tests/behaviour/steps/inspectcli.py [hunk 5]
```diff
@@ -1,12 +1,37 @@
+import pandas as pd
 from difflib import ndiff
 from pathlib import Path
 
 from behave import *
+from pandas.util.testing import assert_frame_equal
+from more_itertools import first
```
Comment (robons) [inline tests/behaviour/steps/inspectcli.py:7]: We already have a function called `first` defined [here](https://github.com/GSS-Cogs/csvcubed/blob/main/src/csvcubed/utils/iterables.py). Can you use that instead so we don't add any new dependencies, even if they are only in testing.

File: tests/behaviour/steps/inspectcli.py [hunk 6]
```diff
@@ -59,9 +90,11 @@ def step_impl(context):
 def step_impl(context):
     metadata_printer = MetadataPrinter(
         context.csvw_type,
+        context.csvw_shape,
         context.csvw_metadata_rdf_graph,
         context.csvw_metadata_json_path,
     )
+    # TODO: Remove below once all the tests are updated to not match strings
```
Comment (robons) [inline tests/behaviour/steps/inspectcli.py:97]: Do we need this still?

File: tests/behaviour/steps/inspectcli.py [hunk 7]
```diff
@@ -81,12 +114,26 @@ def step_impl(context):
         and context.dataset_observations_info_printable
         and context.dataset_val_counts_by_measure_unit_info_printable
     )
+    # TODO: Remove above once all the tests are updated to not match strings
```
Comment (robons) [inline tests/behaviour/steps/inspectcli.py:117]: And again?

File: tests/behaviour/steps/inspectcli.py [hunk 8]
```diff
@@ -140,6 +187,7 @@ def step_impl(context):
 
 @Then("the Dataset Information Printable should be")
 def step_impl(context):
+    print(context.dataset_observations_info_printable)
```
Comment (robons) [inline tests/behaviour/steps/inspectcli.py:190]: Can you remove this?

File: tests/behaviour/steps/inspectcli.py [hunk 9]
```diff
@@ -171,3 +219,341 @@ def step_impl(context, error_message: str):
...
+    assert (
+        len(result_catalog_metadata.themes) == 1
+        and result_catalog_metadata.themes[0] == ""
+    )
+    assert (
+        len(result_catalog_metadata.keywords) == 1
+        and result_catalog_metadata.keywords[0] == ""
+    )
+    assert result_catalog_metadata.contact_point == "None"
+    assert result_catalog_metadata.identifier == "qb-id-10004"
```
Comment (robons) [inline tests/behaviour/steps/inspectcli.py:255]: Should this be an input to the step definition? 

Can you either reword the step definition so it takes all the values which might change as parameters, or make the step name so specific no one else will attempt to reuse it?

File: tests/behaviour/steps/inspectcli.py [hunk 10]
```diff
@@ -171,3 +219,341 @@ def step_impl(context, error_message: str):
...
+    result_qube_components: QubeComponentsResult = context.result_qube_components
+    assert result_qube_components is not None
+
+    components = result_qube_components.qube_components
+    assert len(components) == 5
+
+    component = get_dsd_component_by_property_url(
+        components, "qb-id-10004.csv#dimension/some-dimension"
+    )
+    assert_dsd_component_equal(
```
Comment (robons) [inline tests/behaviour/steps/inspectcli.py:273]: These tests are so specific I'm not sure they really fit into the behaviour testing framework. 

Could you convert them into unit tests? Unless we absolutely need any of the docker-based functionality.

File: tests/unit/utils/sparqlhandler/test_sparqlmanager.py [hunk 11]
```diff
@@ -159,32 +241,203 @@ def test_select_csvw_catalog_metadata_for_codelist():
     assert result.identifier == "Alcohol Content"
 
 
-def test_select_csvw_dsd_dataset():
+def test_select_csvw_dsd_dataset_for_pivoted_multi_measure_data_set():
```
Comment (robons) [inline tests/unit/utils/sparqlhandler/test_sparqlmanager.py:244]: Do we need to test this for the old-style single measure pivoted input? The inspect functionality should still be capable of querying and interpreting it, even if the corresponding version of csvcubed can't output it anymore.

File: src/csvcubed/cli/inspect/metadatainputvalidator.py [hunk 12]
```diff
@@ -41,20 +43,17 @@ class MetadataValidator:
...
+        Detects the validity, type and shape of the csvw.
         """
-        metadata_type = self._detect_type()
-        return (
-            metadata_type == CSVWType.QbDataSet or metadata_type == CSVWType.CodeList,
-            metadata_type,
-        )
+        csvw_type = self._detect_type()
+        csvw_shape = self._detect_shape() if csvw_type == CSVWType.QbDataSet else None
+        validity = csvw_type == CSVWType.QbDataSet or csvw_type == CSVWType.CodeList
```
[the remark this answers is not in the document]
Comment (GDonRanasinghe) [reply src/csvcubed/cli/inspect/metadatainputvalidator.py]: You are correct. We have updated it to throw an exception.

File: src/csvcubed/cli/inspect/metadatainputvalidator.py [hunk 13]
```diff
@@ -78,3 +77,22 @@ def _detect_type(self) -> CSVWType:
...
         else:
             return CSVWType.Other
+
+    def _detect_shape(self) -> CSVWShape:
+        """
+        Given a metadata validator as input, returns the shape of the cube that metadata describes (Pivoted or Standard).
+        """
+        all_pivoted = True
+        all_standard_shape = True
+        for measure in self.is_pivoted_measures:
```
[the remark this answers is not in the document]
Comment (GDonRanasinghe) [reply src/csvcubed/cli/inspect/metadatainputvalidator.py:88]: @robons We have updated this as follows:
1. The `self.is_pivoted_measures` is not a class property now.
2. The sparql for detecting pivoted measures is run in the inspect.py class. This is similar to how the other sparqls are being run from this class.

File: tests/behaviour/steps/inspectcli.py (hunk 6 above)
Comment (GDonRanasinghe) [reply tests/behaviour/steps/inspectcli.py:77]: Yes, this is related to the old text matching style behave tests.

File: tests/behaviour/steps/inspectcli.py (hunk 7 above)
Comment (GDonRanasinghe) [reply tests/behaviour/steps/inspectcli.py:97]: Yes, this is related to the old text matching style behave tests.

File: tests/behaviour/steps/inspectcli.py (hunk 10 above)
Comment (GDonRanasinghe) [reply tests/behaviour/steps/inspectcli.py:249]: We already have unit tests for this. If we are moving away from the test based approach for behave tests, we should test the objects? DSD components are objects.

File: src/csvcubed/models/sparqlresults.py [hunk 14]
```diff
@@ -358,7 +358,7 @@ def _map_qube_component_sparql_result(
             str(result_dict["componentPropertyType"])
         ),
         csv_col_title=none_or_map(result_dict.get("csvColumnTitle"), str) or "",
-        observation_value_column_titles="",
+        observation_value_column_titles="", # This value is popuated after runnning the SELECT_OBS_VAL_FOR_DSD_COMPONENT_PROPERTIES sparql query. The _map_obs_val_for_dsd_component_properties_results will set the value for this property.
```
Comment (robons) [inline src/csvcubed/models/sparqlresults.py:361]: Do you have the black formatter correctly installed and functioning here? It shouldn't allow a line this long.

File: src/csvcubed/utils/sparql_handler/sparqlmanager.py [hunk 15]
```diff
@@ -222,7 +210,7 @@ def select_csvw_dsd_dataset_label_and_dsd_def_uri(
 
 
 def select_csvw_dsd_qube_components(
-    csvw_shape: Optional[CSVWShape],
+    cube_shape: Optional[CubeShape],
```
Comment (robons) [inline src/csvcubed/utils/sparql_handler/sparqlmanager.py:213]: I don't think the cube shape should really be optional at this point, it should require a value.

File: tests/unit/utils/sparqlhandler/test_sparqlmanager.py [hunk 16]
```diff
@@ -240,6 +239,32 @@ def test_select_csvw_catalog_metadata_for_codelist():
     assert result.contact_point == "None"
     assert result.identifier == "Alcohol Content"
 
+def test_select_csvw_dsd_dataset_for_old_style_data_set():
```
Comment (robons) [inline tests/unit/utils/sparqlhandler/test_sparqlmanager.py:242]: Can you make it clearer exactly what this test is doing. An `old style` data set only makes sense with the current context we have. It's going to be very difficult to understand what an old style data set is when a developer comes back and tries to understand this in future.

File: tests/unit/utils/sparqlhandler/test_sparqlmanager.py [hunk 17]
```diff
@@ -240,6 +239,32 @@ def test_select_csvw_catalog_metadata_for_codelist():
...
+        csvw_metadata_rdf_graph
+    )
+    component_result: QubeComponentsResult = select_csvw_dsd_qube_components(
+        CubeShape.Standard, csvw_metadata_rdf_graph, result.dsd_uri, csvw_metadata_json_path
+    )
+    components = component_result.qube_components
+
+    assert result.dataset_label == "Alcohol Bulletin"
+    assert len(components) == 17
+    assert (
```
Comment (robons) [inline tests/unit/utils/sparqlhandler/test_sparqlmanager.py:260]: As we've discussed previously, can you make sure you're not relying on the potentially random order in which the components come back?

**Review by robons — APPROVED** [review]
I'm approving this PR so we can get it out the way, but can you take a look at the new comments I've left and address them in your next PR?

**Pull request merged**